File: cpp_version/dynamic_programming/fib/fib.cpp
/*
fib
Write a function fib that takes in a number argument, n, and
returns the n-th number of the Fibonacci sequence.
The 0-th number of the sequence is 0.
The 1-st number of the sequence is 1.
To generate further numbers of the sequence, calculate
the sum of previous two numbers.
Solve this recursively.

*/
#include <iostream>
#include <unordered_map>
using namespace std;
// brute force method 0(n^2) exponential time via branch ^ depth (number of return branches ^ n)_

// int fib(int num) {
//     if (num == 0) {
//         return 0;
//     }
//     if (num == 1) {
//         return 1;
//     }
//     return fib(num - 2) + fib(num - 1);
// }

//optimized method using memoization

int fib(int num, unordered_map<int, int> &u_map) {

    return u_map[num];
}

int main() {
    std::unordered_map<int,int>u_map;
    cout << fib(0,u_map) << endl;  // 0
    cout << fib(1,u_map) << endl;  // 1
    cout << fib(2,u_map) << endl;  // 1
    cout << fib(3,u_map) << endl;  // 2
    cout << fib(4,u_map) << endl;  // 3
    cout << fib(5,u_map) << endl;  // 5
    cout << fib(35,u_map) << endl; // 9227465
    cout << fib(46,u_map) << endl; // 1836311903
}

File: java_version/dynamic_programming/fib/fib.java
package dynamic_programming.fib;

import java.util.HashMap;

/*
fib
Write a function fib that takes in a number argument, n, and
returns the n-th number of the Fibonacci sequence.
The 0-th number of the sequence is 0.
The 1-st number of the sequence is 1.
To generate further numbers of the sequence, calculate
the sum of previous two numbers.
Solve this recursively.

*/
public class fib {

    public static int Fibonacci(int n, HashMap<Integer, Integer> hash) {

        return hash.get(n);
    }

    public static void main(String[] args) {
        HashMap<Integer, Integer> hash = new HashMap<>();
        System.out.println(Fibonacci(0, hash)); // 0
        System.out.println(Fibonacci(1, hash)); // 1
        System.out.println(Fibonacci(2, hash)); // 1
        System.out.println(Fibonacci(3, hash)); // 2
        System.out.println(Fibonacci(4, hash)); // 3
        System.out.println(Fibonacci(5, hash)); // 5
        System.out.println(Fibonacci(35, hash)); // 9227465
        System.out.println(Fibonacci(46, hash)); // 1836311903
    }
}